beni/__init__.py

```diff
@@ -79,8 +110,32 @@ def generate_environment(
     }
 
 
-def main() -> None:
-    args = parser.parse_args()
+def extras_to_install(c: flit_core.inifile.LoadedConfig, deps: Deps, extras: typing.Sequence[str]) -> typing.Set[str]:
+    to_install = set(extras)
+    if deps is Deps.all or 'all' in to_install:
+        to_install |= set(c.reqs_by_extra.keys())
+    elif deps is Deps.develop:
+        to_install |= {'dev', 'doc', 'test'}
+    return to_install
+
+
+def is_in_extras(req: packaging.requirements.Requirement, extras: typing.Set[str]):
+    if not req.marker:
+        return True
+    for extra in extras:
+        if req.marker.evaluate(dict(extra=extra)):
+            return True
+    return False
+
+
+def main(argv: typing.Optional[typing.Sequence[str]] = None) -> None:
+    if argv is None:
+        argv = sys.argv[1:]
+    args = parser.parse_args(argv)
+    if args.deps is Deps.none and args.extras:
+        raise argparse.ArgumentError(extras_action, "Can’t specify extras together with --deps=none")
+    if args.deps is Deps.all and args.extras:
+        args.deps = Deps.production  # the default is all so override if extras are specified
```

Comment:
This makes it behave differently from flit. I just assume someone specifying `beni --extras=foo` means to install just the `foo` extra, while flit needs you to specify `flit --deps=production --extras=foo` for that.

Reply:
Ah thanks for pointing this out.

I hadn't looked that deeply into how flit has you specify the `--deps` and the `--extras` before, even reading it now it is a bit subtle for me.

So I wonder if it makes sense to either:

1. Copy flit's behavior exactly, so we can direct folks to their docs to understand the semantics
2. See if we can cut the two args down to one, by defaulting to install all extras if no arg is passed, and only have an `--extras` arg, which takes a list of extras and only installs those. 

What do you think of that? I am just considering that if it takes me a while to try to comprehend how these two args mesh it might also be difficult for others to do the same.

But it's not a deal breaker for me, if you think how you have it now makes more sense!

Reply:
I like the way flit does it, I understand `--deps` as shortcut for which requirements to install:

- `none` means “don’t install *any* deps”. It doesn’t make much sense for us, maybe I should remove it
- `production` means “just the base deps, no extras”
- `development` means “all extras needed for building the `doc`s, running the `test`s, and whatever you specified in `dev`”
- `all` is obvious, and also the default, which means that `--deps=all --extras=...` means just `--deps=all` in flit.

Maybe the default should be a special value that is documented as “all extra dependencies if `--extras` is not specified else only those”

Reply:
> Maybe the default should be a special value that is documented as “all extra dependencies if --extras is not specified else only those”

That sounds good.

Yea they are both nice and make sense separately, the only non-obvious choice to me is what happens if a user specifies both of them.

What about this:

>  If `--extras` is a non empty list, only those extras are included and any arg passed to `--deps` is ignored.

Reply:
I think that makes it strictly less useful and behave very differently from flit. I now made it so:

- `--extras=foo,bar` means “install deps for the extras foo and bar”
- `--deps=develop --extras=foo` means “install the usual development extras dev, doc, and test, and also the extra foo”
- `--deps=production` means “don’t install any extra dependencies”